cpp/tests/io/algo/scan_select_if.cuh

```diff
@@ -0,0 +1,432 @@
+  typename Policy::ScanOperator scan_op,
+  typename Policy::PredOperator pred_op,
+  typename Policy::ItemsTileState items_tile_state,
+  typename Policy::IndexTileState index_tile_state,
+  bool do_scatter,
+  uint32_t num_tiles,
+  uint32_t start_tile)
+{
+  auto agent_instance = agent<Policy>{
+    d_input,
+    d_output_count,
+    d_output,
+    num_items,
+    scan_op,
+    pred_op,
+    items_tile_state,
+    index_tile_state  //
+  };
+
+  agent_instance.consume_range(do_scatter, num_tiles, start_tile);
+}
+
+// ===== Policy ====================================================================================
+
+template <typename InputIterator_,
+          typename OutputCountIterator_,
+          typename OutputIterator_,
+          typename ScanOperator_,
+          typename PredicateOperator_>
+struct policy {
+  enum : uint32_t {
+    THREADS_PER_INIT_BLOCK = 128,
+    THREADS_PER_BLOCK      = 128,
+    ITEMS_PER_THREAD       = 16,
+    ITEMS_PER_TILE         = ITEMS_PER_THREAD * THREADS_PER_BLOCK,
+  };
+
+  using InputIterator       = InputIterator_;
+  using OutputIterator      = OutputIterator_;
+  using OutputCountIterator = OutputCountIterator_;
+
+  using ScanOperator = ScanOperator_;
+  using PredOperator = PredicateOperator_;
+
+  using Offset      = typename std::iterator_traits<InputIterator>::difference_type;
+  using Input       = typename std::iterator_traits<InputIterator>::value_type;
+  using OutputCount = typename std::iterator_traits<OutputCountIterator>::value_type;
+  using OutputValue = typename std::iterator_traits<OutputIterator>::value_type;
+
+  // Items Load
+
+  using ItemsBlockLoad = cub::BlockLoad<  //
+    Input,
+    THREADS_PER_BLOCK,
+    ITEMS_PER_THREAD,
+    cub::BlockLoadAlgorithm::BLOCK_LOAD_DIRECT>;
+
+  // Items Scan
+
+  using ItemsTileState      = cub::ScanTileState<Input>;
+  using ItemsPrefixCallback = cub::TilePrefixCallbackOp<Input, ScanOperator, ItemsTileState>;
+
+  using ItemsBlockScan = cub::BlockScan<  //
+    Input,
+    THREADS_PER_BLOCK,
+    cub::BlockScanAlgorithm::BLOCK_SCAN_WARP_SCANS>;
+
+  // Index Scan
+
+  using IndexTileState      = cub::ScanTileState<uint32_t>;
+  using IndexPrefixCallback = cub::TilePrefixCallbackOp<uint32_t, cub::Sum, IndexTileState>;
+  using IndexBlockScan      = cub::BlockScan<  //
+    uint32_t,
+    THREADS_PER_BLOCK,
+    cub::BlockScanAlgorithm::BLOCK_SCAN_WARP_SCANS>;
+};
+
+// ===== Entry =====================================================================================
+
+template <typename InputIterator,
+          typename OutputCountIterator,
+          typename OutputIterator,
+          typename ScanOperator,
+          typename PredOperator>
+void scan_select_if(  //
+  void* d_temp_storage,
+  size_t& temp_storage_bytes,
+  InputIterator d_in,
+  OutputCountIterator d_count_out,
+  OutputIterator d_out,
+  uint32_t num_items,
+  ScanOperator scan_op,
+  PredOperator pred_op,
+  bool do_initialize,
+  bool do_scatter,
```

Comment:
I agree. I was trying to figure out a clean way to solve this, but this seemed the cleanest. The problem is that the code paths are nearly identical. throwing in a couple of bools to enable/disable two sections seemed clean from the implementation standpoint, but utilizing the function requires three calls.

We could add an "is tile state initialized" flag to the temp memory, create the temp memory internally, initialize it to "tile state not initialized", remove `bool do_initialize`, and return the temp memory such that it can be passed to the second and third call.

We could specialize the kernel to accept "void" as an OutputIterator, and disable scatter via that, but that would increase compilation time (trivially?) and make the implementation more complicated.

Or we can have two bools, both of which have no effect on the other (all four combinations are valid, though initializing on the scatter phase is practically useless), and only enable/disable an independent section of code.

Do you have an idea? I'd much rather have a better solution.